Added `api/collection/{tree}` flag to remove other users' collections from results

markbastian opened the pull request:
PR for #28199

markbastian commented on src/metabase/api/collection.clj:
```diff
@@ -36,46 +36,90 @@
+   :where     (cond-> (into [:and] (or and-clauses [true]))
+                      exclude-other-user-collections? (conj [:= :a.visible true]))
+   :order-by  [[:%lower.name :asc]]})
+
```
Note that one design consideration in this was to branch on exclude-other-user-collections? (since it needs the join logic) or just have one query that may not use the join. I opted for the single query as it makes the code more cohesive.

qnkhuat commented on test/metabase/api/collection_test.clj:
```diff
@@ -486,63 +519,63 @@
-                     (assoc-in [1 :fully_parametrized] true))
-                 (mt/boolean-ids-and-timestamps
-                  (:data (mt/user-http-request :rasta :get 200 (str "collection/" (u/the-id collection) "/items"))))))))
+                                          (is (= (-> (mapv default-item [{:name "Acme Products", :model "pulse", :entity_id true}
```
pretty sure the previous indentation is right.
The `with-some-children-of-collection` has a `style/indent=1`

markbastian replied:
fixed

qnkhuat commented on test/metabase/api/collection_test.clj:
```diff
@@ -1268,60 +1301,60 @@
   (testing "POST /api/collection"
     (testing "\ntest that we can create a new collection"
       (mt/with-model-cleanup [Collection]
```
same, I think previous indentation was correct

markbastian replied:
Fixed

qnkhuat commented on test/metabase/api/collection_test.clj:
```diff
@@ -1488,53 +1521,53 @@
           (testing "Grant perms for A and B. Moving A into C should fail because we need perms for C"
             ;; (collections with readwrite perms marked below with a `*`)
             ;; A* -> B* ==> C -> A -> B
             (mt/with-non-admin-groups-no-root-collection-perms
```
The previous indentation is correct  https://guide.clojure.style/#body-indentation

markbastian replied:
Fixed

ranquild commented:
@markbastian just tried your PR and it seems that I need `personal_owner_id` in `effective_ancestors` for each collection in `/api/collection/:id` to make everything work. WDYT?

markbastian commented:
> @markbastian just tried your PR and it seems that I need `personal_owner_id` for each collection in `effective_ancestors` in `/api/collection/:id` to make everything work. WDYT?

Yep, done.

markbastian committed:
Local S&M

markbastian committed:
Merge branch 'master' into flag-to-hide-colls-28199

ranquild reviewed (changes requested):
I'll block the PR until we figure FE stuff.

ranquild committed:
Do not load personal collections on home page

ranquild committed:
Do not load personal collections on home page

markbastian committed:
Merge branch 'master' into flag-to-hide-colls-28199

ranquild reviewed (approved):
I didn't initially want to go with this change but given the pressure it seems to be what we can ship shortly. This will only help with the initial load but there would be still a huge delay when creating a question from another question.

markbastian committed:
Merge branch 'flag-to-hide-colls-28199' into flag-to-hide-colls-28199-fe

ranquild committed:
[WIP] Do not load personal collections on home page (#28576)

Do not load personal collections on home page.

markbastian committed:
Merge remote-tracking branch 'origin/flag-to-hide-colls-28199' into flag-to-hide-colls-28199

qnkhuat commented on src/metabase/api/collection.clj:
```diff
@@ -45,17 +45,54 @@
+         (if (or (nil? personal_owner_id) (= user-id personal_owner_id))
+           (recur r (conj keep-roots id) skip-roots (conj res c) indeterminates)
+           (recur r keep-roots (conj skip-roots id) res indeterminates))
+         (let [[_ root] (str/split location #"/")
```
we have a helper fn to explode the location in `metabase.models.collection/location-path->ids` maybe it's better to use that fn so we don't have to do the same thing of split->parse over and over.

markbastian replied:
Done

qnkhuat commented on src/metabase/api/collection.clj:
```diff
@@ -45,17 +45,54 @@
+          skip-roots     #{}
+          res            []
+          indeterminates []]
+     (if c
```
there are lots of conditions here and it takes time to follow this, maybe you could add some comments explaining what each branch does?

markbastian replied:
Done

qnkhuat commented on src/metabase/api/collection.clj:
```diff
@@ -45,17 +45,54 @@
+               (let [[_ root] (str/split location #"/")]
+                 (keep-roots (parse-long root)))))
+            (into res)))))
+  ([collections]
```
I think the consensus is having low arity defined first in a multi-arity fn, so maybe move this up?

markbastian replied:
Done

qnkhuat commented on test/metabase/api/collection_test.clj:
```diff
@@ -117,16 +149,16 @@
-                                (or (#{"Our analytics" "Collection 1" "Collection 2"} collection-name)
-                                    (str/includes? collection-name "Personal Collection"))))
-                      (map :name)))))))
+       (mt/with-temp* [Collection [collection-1 {:name "Collection 1"}]
```
the original indentation was correct

markbastian replied:
Fixed

qnkhuat commented on test/metabase/api/collection_test.clj:
```diff
@@ -103,10 +103,42 @@
+    (testing "You should only see your collection and public collections"
+      (let [admin-user-id  (u/the-id (test.users/fetch-user :crowberto))
+            crowberto-root (db/select-one Collection :personal_owner_id admin-user-id)]
+        (mt/with-temp* [Collection [collection]
```
nitty: we're moving to toucan 2 and it's recommended to use `t2.with-temp/with-temp` now. we should do this with new tests because it takes quite an effort to rewrite between the 2 versions

markbastian replied:
Done

qnkhuat reviewed (changes requested):
Sorry it takes too long for you to get this reviewed.

The logic works but we haven't added tests for the `GET /api/collection/tree` endpoint.

Others than that I have some comments about styling, feel free to address them if you find it's appropriate.

markbastian committed:
Merge branch 'master' into flag-to-hide-colls-28199

markbastian committed:
Using location-path->ids

markbastian commented on src/metabase/api/collection.clj:
```diff
@@ -45,17 +45,54 @@
+  `?exclude-other-user-collections=true`.
+  "
+  [archived exclude-other-user-collections namespace]
   {archived  (s/maybe su/BooleanString)
```
Done

markbastian committed:
formatting

markbastian committed:
indentation

markbastian committed:
formatting + t2

markbastian committed:
Merge branch 'master' into flag-to-hide-colls-28199

markbastian committed:
/tree test for exclude-other-user-collections flag

markbastian committed:
Added and fixed test

markbastian committed:
Merge branch 'master' into flag-to-hide-colls-28199

dpsutton commented on src/metabase/api/collection.clj:
```diff
@@ -45,17 +45,68 @@
+   collections appear before children, but will catch any mis-ordered entries and
+   do a final pass on those if needed."
+  ([collections]
+   (remove-other-users-personal-collections (:id @api/*current-user*) collections))
```
in the future you can use `metabase.api.common/*current-user-id*` or feel free to change it now if you like.

markbastian replied:
Done

markbastian committed:
Updated to use current user id

markbastian committed:
Merge branch 'master' into flag-to-hide-colls-28199

metamben commented on src/metabase/api/collection.clj:
```diff
@@ -45,17 +45,68 @@
+       ;; private shared collection was made public (the parent is private, so no root node, but the child is visible).
+       ;; At this stage, we just do a final sweep to remove anything for which we know the root collection belongs to
+       ;; another user.
+       (->> indeterminates
+            (remove
+             (fn [{:keys [location]}]
+               (let [[root] (collection/location-path->ids location)]
+                 (skip-roots root))))
+            ;; Put anything that remains in the result and return
+            (into res))))))
```
```suggestion
       ;; Put anything that remains in the result and return
       (into res
             (remove (comp skip-roots first collection/location-path->ids :location))
             indeterminates)))))
```

metamben commented on src/metabase/api/collection.clj:
```diff
@@ -45,17 +45,68 @@
-  {archived  (s/maybe su/BooleanString)
+  `?archived=true`.
+
+  By default, admin users will see all collections. To hide other user's collections pass in\n
+  `?exclude-other-user-collections=true`.
+  "
```
```suggestion
  By default, admin users will see all collections. To hide other user's collections pass in
  `?exclude-other-user-collections=true`."
```

metamben commented on test/metabase/api/collection_test.clj:
```diff
@@ -240,7 +272,43 @@
+                    {:name "Crowberto Corv's Personal Collection", :children [{:name "Admin Child", :children []}]}]
+                   (collection-tree-view ids admin-response-ex)))
+              ;; A non admin only sees their own collections and shared collections no matter what
+              (is (= [{:name "A", :children [{:name "B", :children []}]}
+                      {:name "C", :children []}
+                      {:name "Lucky Pigeon's Personal Collection", :children [{:name "Lucky Child", :children []}]}]
+                     (collection-tree-view ids non-admin-response)
+                     (collection-tree-view ids non-admin-response-ex))))))))
```
Separate `(is ...)` clauses are easier to follow if one of the actual values is unexpected.

metamben commented on test/metabase/api/collection_test.clj:
```diff
@@ -1678,14 +1750,14 @@
           (testing "Should ignore updates to Collections outside of the namespace"
             (let [response (mt/user-http-request :crowberto :put 200 "collection/graph"
                                                  (assoc (graph/graph)
-                                                        :groups {group-id {default-a :write, currency-a :write}}
-                                                        :namespace :currency))]
+                                                   :groups {group-id {default-a :write, currency-a :write}}
+                                                   :namespace :currency))]
```
I thought the original indentation was standard.

metamben commented on test/metabase/api/collection_test.clj:
```diff
@@ -1678,14 +1750,14 @@
           (testing "Should ignore updates to Collections outside of the namespace"
             (let [response (mt/user-http-request :crowberto :put 200 "collection/graph"
                                                  (assoc (graph/graph)
-                                                        :groups {group-id {default-a :write, currency-a :write}}
-                                                        :namespace :currency))]
+                                                   :groups {group-id {default-a :write, currency-a :write}}
+                                                   :namespace :currency))]
               (is (= {"Currency A" "write", "Currency A -> B" "read"}
                      (nice-graph response))))))
 
         (testing "have to be a superuser"
           (is (= "You don't have permissions to do that."
                  (mt/user-http-request :rasta :put 403 "collection/graph"
                                        (assoc (graph/graph)
-                                              :groups {group-id {default-a :write, currency-a :write}}
-                                              :namespace :currency)))))))))
+                                         :groups {group-id {default-a :write, currency-a :write}}
+                                         :namespace :currency)))))))))
```
I thought the original indentation was standard.

dpsutton commented:
@markbastian I wonder if we can simplify the function a bit more

```clojure
(defn- remove-other-users-personal-collections
  [user-id collections]
  (let [prefixes (into [] (comp (filter :personal_owner_id)
                                (remove (comp #{user-id} :personal_owner_id))
                                (distinct)
                                (map (fn [collection] (format "/%d/" (:id collection)))))
                       collections)
        personal? (fn [{^String location :location}]
                   (some (fn [^String prefix] (.startsWith location prefix))
                         prefixes))]
    (if (seq prefixes)
      (remove personal? collections)
      collections)))
```
This i think is perhaps more clear than your version here. We just enumerate the personal roots of others and check each collection's location again them. Is this more clear?

A quick test

```clojure
collection=> (remove-other-users-personal-collections
               1
               [{:location "/"
                 :name "my personal"
                 :id 1
                 :personal_owner_id 1}
                {:location "/"
                 :name "your personal"
                 :id 2
                 :personal_owner_id 2}
                {:location "/2/" :id 3 :name "personal sub"}
                {:location "/2/3/" :id 4 :name "personal sub sub"}
                {:location "/" :id 5 :name "general"}
                {:location "/5/" :id 5 :name "general sub"}])

({:location "/", :name "my personal", :id 1, :personal_owner_id 1}
 {:location "/", :name "your personal", :id 2, :personal_owner_id 2}
 {:location "/", :id 5, :name "general"}
 {:location "/5/", :id 5, :name "general sub"})
```

markbastian committed:
Update src/metabase/api/collection.clj

Co-authored-by: metamben <103100869+metamben@users.noreply.github.com>

markbastian committed:
Update src/metabase/api/collection.clj

Co-authored-by: metamben <103100869+metamben@users.noreply.github.com>

dpsutton commented on src/metabase/api/collection.clj:
```diff
@@ -64,7 +111,11 @@
-      ;; include Root Collection at beginning or results if archived isn't `true`
+          ;; Remove other users' personal collections
+          (if exclude-other-user-collections?
+            (remove-other-users-personal-collections collections)
```
your overload reaches out to `api/*current-user-id*` in an optional overload. I think you should just pass it in. Makes it obvious when testing to supply it. Otherwise it will get a null user-id if people don't realize it grabs that itself.

markbastian committed:
Split out is clauses

markbastian committed:
ws

markbastian committed:
Removing collection arity on remove-other-users-personal-collections

markbastian committed:
Using simpler version from Dan

markbastian committed:
Merge branch 'master' into flag-to-hide-colls-28199

markbastian commented:
@dpsutton - I liked your first version. I added that in and added in 6000 users again. Did some quit time testing and it's comparable. Let's go with that.

markbastian commented:
A bit of evidence:
SELECT COUNT(*) FROM collection; -- gives 6005

<img width="1109" alt="image" src="https://user-images.githubusercontent.com/8381441/221324901-b74661ea-06ba-4829-8859-5e59ea76e492.png">

And a nice recording of the differences:
![timings](https://user-images.githubusercontent.com/8381441/221325292-42b8d664-efe0-4b02-bc9d-8aa36f2267a1.gif)

markbastian commented:
And here's the snappy desktop with all of those collections eliminated on load.

![fasterdesktop](https://user-images.githubusercontent.com/8381441/221325437-53f5d5af-a302-4400-85f5-ff116462d4b7.gif)

markbastian commented:
A recording of creation of multiple collections and showing the improved performance.

![onemore](https://user-images.githubusercontent.com/8381441/221325869-9f51e820-f4a6-498c-935b-40450d44af4d.gif)

markbastian committed:
Merge branch 'master' into flag-to-hide-colls-28199

qnkhuat reviewed (approved):
:shipit:

dpsutton reviewed (approved):
approving the BE changes. I still think there could be something to grabbing the prefix and checking a set of prefixes rather than looping over each user.

markbastian committed:
O(n) version

markbastian committed:
Merge branch 'master' into flag-to-hide-colls-28199
